File: python/forge/training/muzero_mc/manifest.py
"""Python mirror of ``forge_mc_runner::ModelManifest``.

This module is the Python side of the model hot-reload contract. The
Rust runner (``crates/forge-mc-runner/src/manifest.rs``) and this
module MUST agree on the JSON shape; any drift breaks the
``schema_version`` cross-check at runtime.

Cross-language constants:

- :data:`MANIFEST_SCHEMA_VERSION` — pinned at ``1``; bumping is a
  breaking change (Rust + Python + any other consumer must bump in
  lockstep).
- :data:`ONNX_OPSET_VERSION` — ONNX opset used by the bootstrap and
  trainer exports. Pinned alongside the Rust ``onnxruntime`` version
  so cross-runtime tests catch opset drift.

Atomic write semantics match the Rust side: writes go to a ``.tmp``
sibling and are then renamed into place. The Rust ``HotReloadWatcher``
relies on this so it never observes a half-written manifest.

No hard-coded values: every dimension, hash, and file role flows
through the :class:`ModelManifest` / :class:`ModelManifestFiles`
dataclasses. Call sites construct manifests via :func:`build_manifest`
or by assembling the dataclasses directly — no raw dict literals.
"""

from __future__ import annotations

__all__ = [
    "DEFAULT_BUNDLE_FILENAMES",
    "MANIFEST_FILENAME",
    "MANIFEST_SCHEMA_VERSION",
    "ONNX_OPSET_VERSION",
    "ModelFileEntry",
    "ModelManifest",
    "ModelManifestFiles",
    "build_manifest",
    "load_manifest",
    "save_manifest",
    "sha256_file",
    "utc_now_rfc3339",
]

import contextlib
import hashlib
import json
import logging
import os
import tempfile
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Final

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------
# Cross-language constants
# ---------------------------------------------------------------------

#: Pinned manifest schema version. Mirrors the Rust constant
#: ``forge_mc_runner::manifest::MANIFEST_SCHEMA_VERSION``. Bumping is a
#: breaking change.
MANIFEST_SCHEMA_VERSION: Final[int] = 1

#: Default ONNX opset for exports. Pinned alongside ``onnxruntime``
#: pulled in by the ``minecraft`` optional-deps group. A round-trip CI
#: test should fail if this drifts away from what the Rust loader
#: accepts.
ONNX_OPSET_VERSION: Final[int] = 17

#: Canonical filename the Rust runner watches for. Bootstrap and
#: trainer exports should write to this name in the bundle directory.
MANIFEST_FILENAME: Final[str] = "model_manifest.json"

#: Default per-role ONNX filenames inside a bundle directory. Callers
#: can override on a per-role basis via :class:`ModelManifestFiles`,
#: but defaulting through this dict keeps every export consistent.
DEFAULT_BUNDLE_FILENAMES: Final[dict[str, str]] = {
    "representation": "representation.onnx",
    "dynamics": "dynamics.onnx",
    "prediction": "prediction.onnx",
}

# Read-buffer size for SHA-256 file hashing. Tuned for typical SSD
# block sizes; not exposed as a config because it has no behavioural
# implications.
_HASH_CHUNK_BYTES: Final[int] = 64 * 1024


# ---------------------------------------------------------------------
# Dataclasses
# ---------------------------------------------------------------------


@dataclass(frozen=True)
class ModelFileEntry:
    """A single per-role file entry inside the manifest.

    Attributes:
        path: Path relative to the manifest directory (or absolute).
            Stored as a string for cross-language JSON portability —
            the Rust side reads this as `String`, not `PathBuf`.
        sha256: Hex SHA-256 digest of the file's bytes. The Rust runner
            recomputes this on load and rejects the bundle on mismatch.
    """

    path: str
    sha256: str


@dataclass(frozen=True)
class ModelManifestFiles:
    """The three per-role file entries a MuZero bundle ships."""

    representation: ModelFileEntry
    dynamics: ModelFileEntry
    prediction: ModelFileEntry


@dataclass(frozen=True)
class ModelManifest:
    """Top-level manifest dataclass — mirrors the Rust struct exactly.

    See ``crates/forge-mc-runner/src/manifest.rs`` for the canonical
    schema. Field order, names, and types must stay in lockstep.
    """

    schema_version: int
    version: int
    schema_id: str
    created_at: str
    files: ModelManifestFiles

    # ---- conversion helpers ----

    def to_json_dict(self) -> dict[str, object]:
        """Render as a JSON-serialisable dict whose shape matches the
        Rust ``serde_json`` output.

        Uses ``dataclasses.asdict`` so any future field additions
        propagate automatically.
        """
        return asdict(self)

    @classmethod
    def from_json_dict(cls, data: dict[str, object]) -> ModelManifest:
        """Build a manifest from a parsed JSON dict.

        Raises ``ValueError`` if required keys are missing or have the
        wrong type — these become :class:`ManifestError` at the
        ``load_manifest`` boundary.
        """
        try:
            files_dict = data["files"]
            if not isinstance(files_dict, dict):
                raise ValueError("files: expected object")
            files = ModelManifestFiles(
                representation=_entry_from_dict(files_dict.get("representation")),
                dynamics=_entry_from_dict(files_dict.get("dynamics")),
                prediction=_entry_from_dict(files_dict.get("prediction")),
            )
            return cls(
                schema_version=_require_int(data, "schema_version"),
                version=_require_int(data, "version"),
                schema_id=str(data["schema_id"]),
                created_at=str(data["created_at"]),
                files=files,
            )
        except KeyError as e:
            raise ValueError(f"manifest missing key: {e.args[0]}") from e

    # ---- validation ----

    def validate(self) -> None:
        """Lightweight cross-check that this manifest could be loaded
        by the Rust runner.

        Raises :class:`ManifestError` on the same conditions as
        ``forge_mc_runner::ModelManifest::validate`` so client code
        sees identical failure modes regardless of where the manifest
        was constructed.
        """
        if self.schema_version != MANIFEST_SCHEMA_VERSION:
            raise ManifestError(
                f"manifest schema_version mismatch: expected "
                f"{MANIFEST_SCHEMA_VERSION}, got {self.schema_version}"
            )
        if self.version < 1:
            raise ManifestError("manifest version must be >= 1")
        if not self.schema_id:
            raise ManifestError("schema_id must be non-empty")
        for role in ("representation", "dynamics", "prediction"):
            entry: ModelFileEntry = getattr(self.files, role)
            if not entry.path:
                raise ManifestError(f"files.{role}.path must be non-empty")
            if not entry.sha256:
                raise ManifestError(f"files.{role}.sha256 must be non-empty")


# ---------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------


class ManifestError(Exception):
    """Raised on manifest schema/validation failures."""


# ---------------------------------------------------------------------
# I/O helpers
# ---------------------------------------------------------------------


def utc_now_rfc3339() -> str:
    """Return the current UTC time formatted as RFC 3339 with ``Z``
    suffix — the canonical form the Rust ``chrono::Utc::now().to_rfc3339()``
    produces, modulo subsecond precision which both sides tolerate.
    """
    return (
        datetime.now(timezone.utc)
        .replace(microsecond=0)
        .isoformat()
        .replace("+00:00", "Z")
    )


def sha256_file(path: str | os.PathLike[str]) -> str:
    """Stream-hash a file with SHA-256 and return its hex digest.

    Streaming (rather than ``hashlib.sha256(open(path, 'rb').read())``)
    keeps memory bounded for large ONNX exports.
    """
    p = Path(path)
    h = hashlib.sha256()
    with p.open("rb") as f:
        while True:
            chunk = f.read(_HASH_CHUNK_BYTES)
            if not chunk:
                break
            h.update(chunk)
    return h.hexdigest()


def _entry_from_dict(d: object) -> ModelFileEntry:
    if not isinstance(d, dict):
        raise ValueError("expected file entry object")
    return ModelFileEntry(path=str(d["path"]), sha256=str(d["sha256"]))


def _require_int(data: dict[str, object], key: str) -> int:
    """Pull an integer field out of an untrusted JSON dict.

    Mypy sees the loaded dict as ``dict[str, object]`` and refuses to
    accept ``int(obj)`` on the bare values; this helper narrows the
    type with an explicit isinstance check, producing the same
    ``ValueError`` shape ``from_json_dict`` already promises.
    """
    v = data[key]
    if isinstance(v, bool):
        # JSON has no separate bool type vs int; reject booleans
        # explicitly so we don't silently accept ``true`` where the
        # Rust side expects a number.
        raise ValueError(f"manifest field {key!r} must be an integer (got bool)")
    if isinstance(v, int):
        return v
    if isinstance(v, str) and v.lstrip("-").isdigit():
        return int(v)
    raise ValueError(f"manifest field {key!r} must be an integer (got {type(v).__name__})")


def load_manifest(path: str | os.PathLike[str]) -> ModelManifest:
    """Load and validate a manifest from disk.

    Raises:
        FileNotFoundError: if the file doesn't exist.
        ManifestError: if the JSON parses but fails validation.
        ValueError: if the JSON is malformed.
    """
    p = Path(path)
    with p.open("r", encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, dict):
        raise ValueError(f"manifest root must be a JSON object, got {type(data).__name__}")
    m = ModelManifest.from_json_dict(data)
    m.validate()
    return m


def save_manifest(manifest: ModelManifest, path: str | os.PathLike[str]) -> Path:
    """Validate and atomically write a manifest to ``path``.

    Atomic semantics: the JSON is written to a ``.tmp`` sibling in the
    same directory, fsync'd, then renamed into place. Matches
    ``forge_mc_runner::ModelManifest::save_json`` so the watcher never
    observes a half-written file.
    """
    manifest.validate()
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)

    payload = json.dumps(manifest.to_json_dict(), indent=2, sort_keys=True)
    # Write to a tmp sibling, fsync, rename. ``delete=False`` because
    # we manage the rename ourselves; ``dir=`` keeps the tmp on the
    # same filesystem so ``os.replace`` is atomic on POSIX and Windows.
    fd, tmp_name = tempfile.mkstemp(prefix=".tmp-", suffix=".manifest", dir=str(p.parent))
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            f.write(payload)
            f.flush()
            os.fsync(f.fileno())
        Path(tmp_name).replace(p)
        logger.debug("manifest saved", extra={"path": str(p), "version": manifest.version})
    except Exception:
        # Best-effort cleanup of the orphan tmp file on failure.
        with contextlib.suppress(OSError):
            Path(tmp_name).unlink()
        raise
    return p


def build_manifest(
    *,
    version: int,
    schema_id: str,
    files_dir: str | os.PathLike[str],
    representation_filename: str = DEFAULT_BUNDLE_FILENAMES["representation"],
    dynamics_filename: str = DEFAULT_BUNDLE_FILENAMES["dynamics"],
    prediction_filename: str = DEFAULT_BUNDLE_FILENAMES["prediction"],
    created_at: str | None = None,
) -> ModelManifest:
    """Assemble a :class:`ModelManifest` from three on-disk ONNX files.

    Computes each role's SHA-256 from the file at
    ``files_dir / <role>_filename``. Stores the per-role ``path`` as
    relative-to-``files_dir`` so the manifest is portable when the
    bundle is later moved.

    Args:
        version: Monotonic export counter.
        schema_id: sha256 of (env_id, obs_dim, action_count, action_map,
            reward_config). Must match the env's handshake `schema_id`.
        files_dir: Directory containing the three ONNX files.
        representation_filename: Override for the representation file
            name (defaults to ``representation.onnx``).
        dynamics_filename: Override for the dynamics file name.
        prediction_filename: Override for the prediction file name.
        created_at: Optional RFC 3339 timestamp; defaults to
            :func:`utc_now_rfc3339`.

    Returns:
        A validated :class:`ModelManifest`.
    """
    files_dir_p = Path(files_dir)
    rep_p = files_dir_p / representation_filename
    dyn_p = files_dir_p / dynamics_filename
    pred_p = files_dir_p / prediction_filename
    files = ModelManifestFiles(
        representation=ModelFileEntry(path=representation_filename, sha256=sha256_file(rep_p)),
        dynamics=ModelFileEntry(path=dynamics_filename, sha256=sha256_file(dyn_p)),
        prediction=ModelFileEntry(path=prediction_filename, sha256=sha256_file(pred_p)),
    )
    m = ModelManifest(
        schema_version=MANIFEST_SCHEMA_VERSION,
        version=version,
        schema_id=schema_id,
        created_at=created_at if created_at is not None else utc_now_rfc3339(),
        files=files,
    )
    m.validate()
    return m

File: crates/forge-mc-runner/src/manifest.rs
//! [`ModelManifest`] — the hot-reload swap signal.
//!
//! The manifest lives next to the three ONNX files
//! (`representation.onnx`, `dynamics.onnx`, `prediction.onnx`) and
//! carries:
//!
//! - `schema_version` — pinned constant; bumping is a breaking change.
//! - `version` — strictly monotonic counter; bumped per trainer export.
//! - `schema_id` — sha256 the runner cross-checks against the env's
//!   `Hello.schema_id`. Mismatch = action/reward space drift, hard fail.
//! - `files.<role>.sha256` — content hash recorded by the trainer at
//!   export time. NOTE: [`ModelManifest::validate`] only checks that
//!   this field is non-empty; it does **not** re-hash the file on
//!   disk, so a corrupted bundle is not detected here. Partial-write
//!   protection comes from the atomic rename below, not from this hash.
//!
//! ## Atomic write
//!
//! [`ModelManifest::save_json`] writes to a `.tmp` sibling then renames
//! into place. This matches the trainer-side
//! `python/forge/training/muzero_mc/exporter.py` discipline so the
//! runner never observes a half-written manifest.

use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use tracing::{debug, instrument};

use crate::error::RunnerError;

/// Current `schema_version` value for this crate's manifest.
///
/// Bumping is a breaking change — older readers will refuse the file
/// with [`RunnerError::ManifestSchemaMismatch`].
pub const MANIFEST_SCHEMA_VERSION: u32 = 1;

/// Per-role file entry inside a manifest.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ModelFileEntry {
    /// Path on disk, typically relative to the manifest's directory.
    pub path: String,
    /// sha256 hex digest of the file contents. The runner can recompute
    /// this and reject the bundle on mismatch.
    pub sha256: String,
}

/// The three files a MuZero bundle ships, by role.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ModelManifestFiles {
    /// Observation → latent (the representation network).
    pub representation: ModelFileEntry,
    /// (latent, action) → next latent + reward (the dynamics network).
    pub dynamics: ModelFileEntry,
    /// Latent → policy logits + value (the prediction network).
    pub prediction: ModelFileEntry,
}

/// A model bundle manifest. One file per trained checkpoint export.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ModelManifest {
    /// MUST equal [`MANIFEST_SCHEMA_VERSION`].
    pub schema_version: u32,
    /// Monotonically increasing per-export version counter.
    pub version: u64,
    /// sha256 of canonical (env_id, obs_dim, action_count, action_map,
    /// reward_config) — runner cross-checks against env `Hello.schema_id`.
    pub schema_id: String,
    /// RFC3339 timestamp the trainer set at export.
    pub created_at: String,
    /// The three ONNX file entries.
    pub files: ModelManifestFiles,
}

impl ModelManifest {
    /// Construct a new manifest with the compiled `schema_version`.
    pub fn new(
        version: u64,
        schema_id: impl Into<String>,
        created_at: impl Into<String>,
        files: ModelManifestFiles,
    ) -> Self {
        Self {
            schema_version: MANIFEST_SCHEMA_VERSION,
            version,
            schema_id: schema_id.into(),
            created_at: created_at.into(),
            files,
        }
    }

    /// Cheap invariant check. Doesn't touch the filesystem.
    pub fn validate(&self) -> Result<(), RunnerError> {
        if self.schema_version != MANIFEST_SCHEMA_VERSION {
            return Err(RunnerError::ManifestSchemaMismatch {
                expected: MANIFEST_SCHEMA_VERSION,
                got: self.schema_version,
            });
        }
        if self.version == 0 {
            return Err(RunnerError::InvalidManifest(
                "version must be >= 1 (0 reserved for 'pre-bootstrap')".into(),
            ));
        }
        if self.schema_id.is_empty() {
            return Err(RunnerError::InvalidManifest(
                "schema_id must be non-empty".into(),
            ));
        }
        for (role, entry) in [
            ("representation", &self.files.representation),
            ("dynamics", &self.files.dynamics),
            ("prediction", &self.files.prediction),
        ] {
            if entry.path.is_empty() {
                return Err(RunnerError::InvalidManifest(format!(
                    "files.{role}.path must be non-empty"
                )));
            }
            if entry.sha256.is_empty() {
                return Err(RunnerError::InvalidManifest(format!(
                    "files.{role}.sha256 must be non-empty"
                )));
            }
        }
        Ok(())
    }

    /// Load a manifest JSON file. Validates after parse.
    #[instrument(skip_all, fields(path = %path.as_ref().display()))]
    pub fn load_json(path: impl AsRef<Path>) -> Result<Self, RunnerError> {
        let path = path.as_ref();
        let bytes = fs::read(path).map_err(|e| RunnerError::io(path, e))?;
        let m: ModelManifest = serde_json::from_slice(&bytes)?;
        m.validate()?;
        debug!(version = m.version, "manifest loaded");
        Ok(m)
    }

    /// Save the manifest atomically (tmp file + rename in same dir).
    /// Validates before writing — a failing manifest never reaches disk.
    #[instrument(skip(self), fields(path = %path.as_ref().display(), version = self.version))]
    pub fn save_json(&self, path: impl AsRef<Path>) -> Result<(), RunnerError> {
        self.validate()?;
        let path = path.as_ref();
        let dir = path.parent().ok_or_else(|| {
            RunnerError::InvalidManifest(format!(
                "manifest path has no parent dir: {}",
                path.display()
            ))
        })?;
        // `path.parent()` returns `Some("")` for bare filenames like
        // `"model_manifest.json"`; `create_dir_all("")` errors on Windows,
        // so only create when there's a real directory component.
        if !dir.as_os_str().is_empty() && !dir.exists() {
            fs::create_dir_all(dir).map_err(|e| RunnerError::io(dir, e))?;
        }
        // Use the dotted-tmp pattern so a crash mid-write leaves a hidden
        // sibling, not a malformed manifest the watcher might pick up.
        let tmp: PathBuf = dir.join(format!(
            ".{}.tmp",
            path.file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_else(|| "model_manifest.json".into())
        ));
        let bytes = serde_json::to_vec_pretty(self)?;
        fs::write(&tmp, &bytes).map_err(|e| RunnerError::io(&tmp, e))?;
        fs::rename(&tmp, path).map_err(|e| RunnerError::io(path, e))?;
        debug!("manifest written");
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Build a syntactically valid manifest for tests.
    fn sample_manifest(version: u64, schema_id: &str) -> ModelManifest {
        ModelManifest::new(
            version,
            schema_id,
            "2026-05-17T00:00:00Z",
            ModelManifestFiles {
                representation: ModelFileEntry {
                    path: "representation.onnx".into(),
                    sha256: "a".repeat(64),
                },
                dynamics: ModelFileEntry {
                    path: "dynamics.onnx".into(),
                    sha256: "b".repeat(64),
                },
                prediction: ModelFileEntry {
                    path: "prediction.onnx".into(),
                    sha256: "c".repeat(64),
                },
            },
        )
    }

    #[test]
    fn new_uses_pinned_schema_version() {
        let m = sample_manifest(1, "abc");
        assert_eq!(m.schema_version, MANIFEST_SCHEMA_VERSION);
    }

    #[test]
    fn validate_accepts_well_formed_manifest() {
        let m = sample_manifest(3, "abc");
        m.validate().unwrap();
    }

    #[test]
    fn validate_rejects_schema_version_drift() {
        let mut m = sample_manifest(1, "abc");
        m.schema_version = 99;
        let err = m.validate().unwrap_err();
        assert!(matches!(err, RunnerError::ManifestSchemaMismatch { .. }));
    }

    #[test]
    fn validate_rejects_version_zero() {
        let m = sample_manifest(0, "abc");
        let err = m.validate().unwrap_err();
        assert!(matches!(err, RunnerError::InvalidManifest(_)));
    }

    #[test]
    fn validate_rejects_empty_schema_id() {
        let m = sample_manifest(1, "");
        let err = m.validate().unwrap_err();
        assert!(matches!(err, RunnerError::InvalidManifest(_)));
    }

    #[test]
    fn validate_rejects_empty_file_path() {
        let mut m = sample_manifest(1, "abc");
        m.files.representation.path = String::new();
        let err = m.validate().unwrap_err();
        match err {
            RunnerError::InvalidManifest(msg) => assert!(msg.contains("representation")),
            other => panic!("expected InvalidManifest, got {other:?}"),
        }
    }

    #[test]
    fn validate_rejects_empty_file_sha() {
        let mut m = sample_manifest(1, "abc");
        m.files.dynamics.sha256 = String::new();
        let err = m.validate().unwrap_err();
        match err {
            RunnerError::InvalidManifest(msg) => assert!(msg.contains("dynamics")),
            other => panic!("expected InvalidManifest, got {other:?}"),
        }
    }

    #[test]
    fn save_then_load_roundtrip() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("model_manifest.json");
        let m = sample_manifest(2, "abcd");
        m.save_json(&path).unwrap();
        let back = ModelManifest::load_json(&path).unwrap();
        assert_eq!(m, back);
    }

    /// The .tmp sibling must not remain after a successful save.
    /// This is what guarantees atomic visibility — the watcher only ever
    /// sees the final `model_manifest.json`.
    #[test]
    fn save_cleans_up_tmp_after_success() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("model_manifest.json");
        let m = sample_manifest(1, "abc");
        m.save_json(&path).unwrap();
        let leftover_tmp = tmp.path().join(".model_manifest.json.tmp");
        assert!(!leftover_tmp.exists());
        assert!(path.exists());
    }

    /// Loading a missing file must produce a path-bearing Io error
    /// (not a generic 'JSON parse failed').
    #[test]
    fn load_missing_file_produces_io_error_with_path() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("nope.json");
        let err = ModelManifest::load_json(&path).unwrap_err();
        match err {
            RunnerError::Io { path: p, .. } => assert_eq!(p, path),
            other => panic!("expected Io, got {other:?}"),
        }
    }

    /// Loading a syntactically invalid JSON file must produce a Json error,
    /// not Io (the file *was* readable; parse is what failed).
    #[test]
    fn load_invalid_json_produces_json_error() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("bad.json");
        std::fs::write(&path, b"{not json").unwrap();
        let err = ModelManifest::load_json(&path).unwrap_err();
        assert!(matches!(err, RunnerError::Json(_)));
    }

    /// Loading a file that parses but violates invariants surfaces
    /// the same validation error as `validate()` directly.
    #[test]
    fn load_file_with_bad_invariants_surfaces_validation_error() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("bad_invariant.json");
        // version=0 → validate() rejects.
        let m = sample_manifest(0, "abc");
        let bytes = serde_json::to_vec_pretty(&m).unwrap();
        std::fs::write(&path, bytes).unwrap();
        let err = ModelManifest::load_json(&path).unwrap_err();
        assert!(matches!(err, RunnerError::InvalidManifest(_)));
    }

    /// Saving an invalid manifest must not leave a file on disk —
    /// validate runs before any IO.
    #[test]
    fn save_invalid_manifest_writes_nothing() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("model_manifest.json");
        let bad = sample_manifest(0, "abc"); // version=0 is invalid
        let _ = bad.save_json(&path).unwrap_err();
        assert!(!path.exists());
    }

    /// `save_json` must create missing intermediate directories so callers
    /// don't have to mkdir `models/` themselves before the first export.
    #[test]
    fn save_json_creates_missing_parent_dirs() {
        let tmp = tempfile::tempdir().unwrap();
        let path = tmp.path().join("a/b/c/model_manifest.json");
        assert!(!path.parent().unwrap().exists());
        let m = sample_manifest(1, "abc");
        m.save_json(&path).unwrap();
        assert!(path.exists());
    }

    /// `Path::new("model_manifest.json").parent()` returns `Some("")`, and
    /// `fs::create_dir_all("")` errors on Windows. The bare-filename case
    /// must still succeed (writes to cwd).
    #[test]
    fn save_json_handles_bare_filename() {
        let tmp = tempfile::tempdir().unwrap();
        // Operate in the tempdir so we don't litter the repo root.
        let prev_cwd = std::env::current_dir().unwrap();
        std::env::set_current_dir(tmp.path()).unwrap();
        let result = sample_manifest(1, "abc").save_json("model_manifest.json");
        let exists = tmp.path().join("model_manifest.json").exists();
        // Restore cwd before any assertion so a failure doesn't leak state.
        std::env::set_current_dir(prev_cwd).unwrap();
        result.unwrap();
        assert!(exists);
    }
}
